### PR TITLE
Add Ender Chest Peeking

### PR DESCRIPTION
**Pull request by davidthehat**
Implements #11 

Only client side, no server behavior.
Includes config toggle.
Does not include translation strings beyond en_us.

### CODE REVIEW

File: src/main/java/de/maxhenkel/peek/data/DataStore.java [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
 import net.minecraft.world.level.block.entity.ShulkerBoxBlockEntity;
 
 import javax.annotation.Nullable;
 
 public class DataStore {
-
     @Nullable
     public static ShulkerBoxBlockEntity lastOpenedShulkerBox;
-
+    public static NonNullList<ItemStack> ecinv = NonNullList.withSize(27, ItemStack.EMPTY);
```
Comment (henkelmax) [inline src/main/java/de/maxhenkel/peek/data/DataStore.java:12]: `enderChestInventory`

Comment (henkelmax) [inline src/main/java/de/maxhenkel/peek/data/DataStore.java:12]: This should be `null` by default to indicate it not being cached. This should also then be annotated with `@Nullable`

File: src/main/java/de/maxhenkel/peek/events/TooltipImageEvents.java [hunk 2]
```diff
@@ -31,11 +32,27 @@ public static Optional<TooltipComponent> getTooltipImage(ItemStack stack, Block
...
 
+    private static Optional<TooltipComponent> getEnderChestTooltipImage() {
+
+        if (!Peek.CONFIG.peekEnderChests.get()) {
+            return null;
+        }
+        //Intended to not check the showEmptyContainers config
+        //This way ender chests are shown as empty only if they have been cached and are actually empty.
+        //When first loading, ender chest rendering is vanilla.
+        if (DataStore.ecinv.stream().allMatch(ItemStack::isEmpty)) {
```
Comment (henkelmax) [inline src/main/java/de/maxhenkel/peek/events/TooltipImageEvents.java:50]: This wouldn't work for ender chests that are empty. Please add the null check here

File: src/main/java/de/maxhenkel/peek/mixin/BlockItemMixin.java [hunk 3]
```diff
@@ -1,17 +1,22 @@
 package de.maxhenkel.peek.mixin;
 
 import de.maxhenkel.peek.Peek;
+import de.maxhenkel.peek.PeekClient;
```
Comment (henkelmax) [inline src/main/java/de/maxhenkel/peek/mixin/BlockItemMixin.java:4]: Why are there new imports if nothing else changed?

File: src/main/java/de/maxhenkel/peek/mixin/ClientPlayNetworkHandlerMixin.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+@Mixin(ClientPacketListener.class)
+public abstract class ClientPlayNetworkHandlerMixin implements ClientGamePacketListener {
+    @Shadow
+    @Final private Minecraft minecraft;
```
Comment (henkelmax) [inline src/main/java/de/maxhenkel/peek/mixin/ClientPlayNetworkHandlerMixin.java:22]: `@Final` should be on its own line

File: src/main/java/de/maxhenkel/peek/mixin/ClientPlayNetworkHandlerMixin.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+    @Inject(method = "handleContainerContent", at = @At("TAIL"))
+    private void onHandleContainerContent(ClientboundContainerSetContentPacket packet, CallbackInfo ci) {
+        if ((this.minecraft.screen instanceof ContainerScreen) && this.minecraft.screen.getTitle().equals(Component.translatable("container.enderchest"))) {
```
Comment (henkelmax) [inline src/main/java/de/maxhenkel/peek/mixin/ClientPlayNetworkHandlerMixin.java:28]: `this` is not necessary here

File: src/main/java/de/maxhenkel/peek/mixin/ClientPlayNetworkHandlerMixin.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+            for (int i = 0; i < packet.getItems().size() - 36; i++) {
+                DataStore.ecinv.set(i, packet.getItems().get(i));
+            }
+        }
+    }
+    @Inject(method = "handleLogin", at = @At("TAIL"))
+    private void onHandleLogin(ClientboundLoginPacket packet, CallbackInfo ci) {
+        DataStore.ecinv = NonNullList.withSize(27, ItemStack.EMPTY);
```
Comment (henkelmax) [inline src/main/java/de/maxhenkel/peek/mixin/ClientPlayNetworkHandlerMixin.java:37]: Use `null` here

File: src/main/java/de/maxhenkel/peek/mixin/ClientPlayNetworkHandlerMixin.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+//        if (minecraft.player != null) {
+//            PlayerEnderChestContainer container = minecraft.player.getEnderChestInventory();
+//            for (int i = 0; i < 27; i++) {
+//                System.out.println("login: "+container.getItem(i));
+//                DataStore.ecinv.set(i, container.getItem(i));
+//            }
+//        }
+
+
+
```
Comment (henkelmax) [inline src/main/java/de/maxhenkel/peek/mixin/ClientPlayNetworkHandlerMixin.java]: Why is this here?

File: src/main/java/de/maxhenkel/peek/config/PeekConfig.java [hunk 8]
```diff
@@ -85,6 +86,11 @@ public PeekConfig(ConfigBuilder builder) {
                 true,
                 "Displays the effect of suspicious stews"
         );
+        peekEnderChests = builder.booleanEntry(
+                "peek_ender_chests",
+                true,
+                "Displays the (cached) contents of ender chests"
```
Comment (henkelmax) [inline src/main/java/de/maxhenkel/peek/config/PeekConfig.java:92]: This should be the same description as for the cloth config description

File: bin/main/peek.mixins.json
Comment (henkelmax) [inline bin/main/peek.mixins.json:1]: You committed the `bin` folder

File: src/main/java/de/maxhenkel/peek/mixin/MultiPlayerGameModeMixin.java [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+@Mixin(MultiPlayerGameMode.class)
+public class MultiPlayerGameModeMixin {
+    @Inject(method = "handleInventoryMouseClick", at = @At("TAIL"))
+    public void onHandleInventoryMouseClick(int i, int j, int k, ClickType clickType, Player player, CallbackInfo ci) {
+        if (DataStore.enderChestOpen) {
+            DataStore.ecinv = player.containerMenu.getItems();
```
Comment (henkelmax) [inline src/main/java/de/maxhenkel/peek/mixin/MultiPlayerGameModeMixin.java:19]: Contents should be also cached when opening and closing the container

File: src/main/java/de/maxhenkel/peek/mixin/ClientPlayNetworkHandlerMixin.java [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    @Inject(method = "handleContainerContent", at = @At("TAIL"))
+    private void onHandleContainerContent(ClientboundContainerSetContentPacket packet, CallbackInfo ci) {
+        if ((this.minecraft.screen instanceof ContainerScreen) && this.minecraft.screen.getTitle().equals(Component.translatable("container.enderchest"))) {
+            for (int i = 0; i < packet.getItems().size() - 36; i++) {
+                DataStore.ecinv.set(i, packet.getItems().get(i));
+            }
+        }
+    }
+    @Inject(method = "handleLogin", at = @At("TAIL"))
+    private void onHandleLogin(ClientboundLoginPacket packet, CallbackInfo ci) {
```
Comment (henkelmax) [inline src/main/java/de/maxhenkel/peek/mixin/ClientPlayNetworkHandlerMixin.java:35]: There is a Fabric API event for this

File: src/main/java/de/maxhenkel/peek/data/DataStore.java [hunk 11]
```diff
@@ -1,12 +1,14 @@
...
+    public static boolean enderChestOpen = false;
```
Comment (henkelmax) [inline src/main/java/de/maxhenkel/peek/data/DataStore.java:13]: I don't know if this is the best solution. There are cases where this can be true without having an ender chest open.

**Review by henkelmax — CHANGES_REQUESTED** [review]

File: src/main/java/de/maxhenkel/peek/mixin/ClientPlayNetworkHandlerMixin.java [hunk 12]
```diff
@@ -23,6 +23,8 @@ public abstract class ClientPlayNetworkHandlerMixin implements ClientGamePacketL
 
     @Inject(method = "handleContainerContent", at = @At("TAIL"))
     private void onHandleContainerContent(ClientboundContainerSetContentPacket packet, CallbackInfo ci) {
+        //Intercept packet on opening enderchest to set contents
+        //Could ignore all but the first, but this is potentially more reliable
```
Comment (henkelmax) [inline src/main/java/de/maxhenkel/peek/mixin/ClientPlayNetworkHandlerMixin.java:27]: Comments should have a space `// `

File: src/main/java/de/maxhenkel/peek/mixin/ClientPlayNetworkHandlerMixin.java [hunk 13]
```diff
@@ -31,17 +33,8 @@ private void onHandleContainerContent(ClientboundContainerSetContentPacket packe
     }
     @Inject(method = "handleLogin", at = @At("TAIL"))
     private void onHandleLogin(ClientboundLoginPacket packet, CallbackInfo ci) {
+        //clear cache on login
```
Comment (henkelmax) [inline src/main/java/de/maxhenkel/peek/mixin/ClientPlayNetworkHandlerMixin.java:36]: This is obvious and doesn't need a comment

File: src/main/java/de/maxhenkel/peek/mixin/MultiPlayerGameModeMixin.java [hunk 14]
```diff
@@ -13,6 +13,8 @@
 public class MultiPlayerGameModeMixin {
     @Inject(method = "handleInventoryMouseClick", at = @At("TAIL"))
     public void onHandleInventoryMouseClick(int i, int j, int k, ClickType clickType, Player player, CallbackInfo ci) {
+        //When the player clicks on an item in the inventory, check if ender chest is open
+        //If it is, update the ender chest cache
```
Comment (henkelmax) [inline src/main/java/de/maxhenkel/peek/mixin/MultiPlayerGameModeMixin.java:17]: This is obvious and doesn't need a comment

**Review by henkelmax — CHANGES_REQUESTED** [review]

File: src/main/java/de/maxhenkel/peek/mixin/ClientPlayNetworkHandlerMixin.java (hunk 7 above)
Comment (davidthehat) [reply src/main/java/de/maxhenkel/peek/mixin/ClientPlayNetworkHandlerMixin.java]: This was removed in the second commit.

File: bin/main/peek.mixins.json
Comment (davidthehat) [reply bin/main/peek.mixins.json:1]: Got it. Should it be added to .gitignore?

File: src/main/java/de/maxhenkel/peek/mixin/MultiPlayerGameModeMixin.java (hunk 9 above)
Comment (davidthehat) [reply src/main/java/de/maxhenkel/peek/mixin/MultiPlayerGameModeMixin.java:19]: Contents are already stored on open using the onHandleContainerContent mixin. Can you clarify why we should also cache on close? An ender chest cannot be modified via any means except a player action. We record the state when the chest is opened for the first time in a session, and we record the state whenever a modification is made.

Comment (henkelmax) [reply src/main/java/de/maxhenkel/peek/mixin/MultiPlayerGameModeMixin.java:19]: It can be changed with other mods.

File: bin/main/peek.mixins.json
Comment (henkelmax) [reply bin/main/peek.mixins.json:1]: You can add it to the .gitignore. But this folder is only generated with eclipse as far as I know

File: src/main/java/de/maxhenkel/peek/data/DataStore.java (hunk 11 above)
Comment (davidthehat) [reply src/main/java/de/maxhenkel/peek/data/DataStore.java:13]: You're right, I clearly didn't check thoroughly enough. The value is de synced when the player moves far enough from the ender chest for it to close automatically.

I really appreciate this note! I realize that I can use the same approach I used in `onHandleContainerContent` in `onHandleInventoryMouseClick` to detect if an ender chest is opened. This eliminates the need to independently track if an ender chest is open. The only way I can see this going wrong is if a container screen has the same translation key as an ender chest. In my opinion, this is acceptable behavior as the client is being told the container screen is an ender chest.

File: bin/main/peek.mixins.json
Comment (davidthehat) [reply bin/main/peek.mixins.json:1]: I am using IntelliJ, I'll add it to the .gitignore.

Comment (henkelmax) [reply bin/main/peek.mixins.json:1]: That's weird...

File: src/main/java/de/maxhenkel/peek/config/PeekConfig.java [hunk 15]
```diff
@@ -89,7 +89,7 @@ public PeekConfig(ConfigBuilder builder) {
         peekEnderChests = builder.booleanEntry(
                 "peek_ender_chests",
                 true,
-                "Displays the (cached) contents of ender chests"
+                "Displays the cached contents of ender chests."
```
Comment (henkelmax) [inline src/main/java/de/maxhenkel/peek/config/PeekConfig.java:92]: Config comments don't have a period

File: src/main/java/de/maxhenkel/peek/mixin/LocalPlayerMixin.java [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+@Mixin(LocalPlayer.class)
+public class LocalPlayerMixin {
+
+    @Shadow @Final protected Minecraft minecraft;
```
Comment (henkelmax) [inline src/main/java/de/maxhenkel/peek/mixin/LocalPlayerMixin.java:17]: Annotations always in a new line

**Review by henkelmax — CHANGES_REQUESTED** [review]

**Comment by henkelmax** [comment]
Theres also still the bin folder present.
PRs should also always have a single commit for readability.

**Comment by davidthehat** [comment]
Apologies for the multiple commits, I missed one of the notes initially.

**Comment by henkelmax** [comment]
You still have multiple commits in the PR. I can't really review everything if there are multiple commits changing stuff.

**Comment by davidthehat** [comment]
Alright, I believe I have fixed it so all changes are in d706bc8b4462136c74ebb8b62df56a12d519e936

**Comment by henkelmax** [comment]
image

Still multiple commits

**Comment by henkelmax** [comment]
Thank you! I will test everything and merge it once I have time. Really appreciate your work!
